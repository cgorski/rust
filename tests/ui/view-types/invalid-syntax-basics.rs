//@ edition:2021
#![feature(view_types)]
#![allow(incomplete_features)]

// NEGATIVE TEST: Basic View Type Syntax Errors
//
// This test covers fundamental syntax errors in view type specifications:
// - Keywords used as field names
// - Empty view specifications
// - Numeric field identifiers (tuple syntax not supported)

// =============================================================================
// ERROR: Keywords cannot be used as field names
// =============================================================================

struct S1 {
    field: i32,
}

impl S1 {
    fn keyword_self(&{mut self} self) {} //~ ERROR expected identifier, found keyword `self`

    fn keyword_super(&{mut super} self) {} //~ ERROR expected identifier, found keyword `super`
}

// =============================================================================
// ERROR: Empty view specification is not allowed
// =============================================================================

struct S2 {
    field: i32,
}

impl S2 {
    fn empty_view(&{} self) {} //~ ERROR expected field name or tuple index
}

// =============================================================================
// NOTE: Tuple fields now supported - this parses successfully
// Tuple struct wrapper used to test tuple field syntax
// =============================================================================

struct TupleWrapper(i32, String);

impl TupleWrapper {
    fn tuple_field(&{mut 0} self) {} // Parses now (tuple fields supported)
}

// =============================================================================
// ERROR: View types require reference types
// =============================================================================

struct S3 {
    field: i32,
}

impl S3 {
    fn owned_value({mut field} self) {} //~ ERROR expected parameter name, found `{`
}

fn main() {}
