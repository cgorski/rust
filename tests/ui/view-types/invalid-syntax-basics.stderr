error: expected identifier, found keyword `self`
  --> $DIR/invalid-syntax-basics.rs:21:27
   |
LL |     fn keyword_self(&{mut self} self) {}
   |                           ^^^^ expected identifier, found keyword

error: expected identifier, found keyword `super`
  --> $DIR/invalid-syntax-basics.rs:23:28
   |
LL |     fn keyword_super(&{mut super} self) {}
   |                            ^^^^^ expected identifier, found keyword

error: expected field name or tuple index
  --> $DIR/invalid-syntax-basics.rs:35:21
   |
LL |     fn empty_view(&{} self) {}
   |                     ^

error: expected parameter name, found `{`
  --> $DIR/invalid-syntax-basics.rs:58:20
   |
LL |     fn owned_value({mut field} self) {}
   |                    ^ expected parameter name

error: aborting due to 4 previous errors

